### ISSUE TITLE
Magick++.h includes assert.h inside namespace.

### ISSUE THREAD

**Issue by plugwash** [ImageMagick/ImageMagick6#95]
### Prerequisites

- [x] I have written a descriptive issue title
- [ ] I have verified that I am using the latest version of ImageMagick

I have tested this with the version currently in Debian sid which is slightly behind the last 6.x release. I have not directly tested with newer versions than that but I have checked the source tree to see if there are any relevant changes and there do not appear to be.

- [x] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

Magick++/Include.h includes the imagemagick headers inside a namespace,
unfortunately the imagemagick headers in turn include the assert.h header which therefore
gets included inside the namespace.

If other code in the same file uses assert.h and Imagemagick is the first thing
to include assert.h this leads to assert functionality being broken for said other
code. For example this causes a build failure of pfstools
https://sourceforge.net/p/pfstools/bugs/48/

### Steps to Reproduce

Build the following code

#include <Magick++.h>
#include <assert.h>
int main() {
    assert(1==2);
}

With "g++ `pkg-config --cflags ImageMagick++`  test.cpp"

### System Configuration

- ImageMagick version:

6.9.11.24

- Environment (Operating system, version and so on):

Debian sid

**Comment by urban-warrior** [comment]
We compiled your source under Fedora and Centos without complaint and on execution we get:
```
a.out: test.cpp:4: int main(): Assertion `1==2' failed.
Abort (core dumped)
```
which appears to be expected behavior.  Perhaps there is a problem in the Debian tool chains.

**Comment by ferdnyc** [comment]
We actually did have an issue with this in our builds as well, which I was similarly unable to reproduce on Fedora. But I did investigate it in our CI builds, before merging a workaround. My full analysis can be found in OpenShot/libopenshot#767 but I'll include the highlights below.

Note that, while this manifested in our code as an interaction with the ZeroMQ headers, that was only due to the fact that they use asserts. (Our own code doesn't.) I saw no evidence the issue had anything to do with the zmq code itself.

> It appears the issue is twofold:
> 
> 1. ImageMagick's `Magick++/Include.h` header namespaces its inclusion of the C headers from `MagickCore` and `MagickWand`:
>     ```c++
>     //
>     // Include ImageMagick headers into namespace "MagickCore". If
>     // MAGICKCORE_IMPLEMENTATION is defined, include ImageMagick development
>     // headers.  This scheme minimizes the possibility of conflict with
>     // user code.
>     //
>     namespace MagickCore
>     {
>     #include <magick/MagickCore.h>
>     #include <wand/MagickWand.h>
>     #undef inline // Remove possible definition from config.h
>     
>     #undef class
>     }
>     ```
>    Those  headers then `#include <assert.h>`.
> 1. ZeroMQ later uses `#include <cassert>`, which depends on `assert.h`. While `assert.h` isn't completely restricted by include guards, a good deal of its functionality is gated on symbols not having already been `#define`d. Since it's a C header and is using the preprocessor, it has no way of knowing that the symbols it already exported are restricted to the `MagickCore` namespace, and it fails to re-define them in the way that ZeroMQ needs.
> 
> I'm not sure why this isn't happening to me on Fedora — it could be that I rarely build in Debug mode, or that I'm building with IM7 while Ubuntu/Debian are still on IM6. Maybe they fixed this somewhere down the line.
> 
> An alternative workaround would be to always ensure that `<zmq.hpp>` is included before `<Magick++.h>`

**Comment by plugwash** [comment]
It turns out there is a Debian-specific aspect to this, specifically a Debian-specific change to assert.h

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=966904#29
https://salsa.debian.org/glibc-team/glibc/-/blob/sid/debian/patches/any/local-asserth-decls.diff 

The patch was introduced to avoid warnings when assert.h was included multiple times.

I still believe that including standard C headers within C++ namespaces is not something programs can reasonably expect to work correctly.

**Comment by ferdnyc** [comment]
The issue also **was not** occurring in our CI builds (I misspoke), running on Bionic and Focal. I only came up in the Launchpad-hosted PPA builds. The build log I posted was from the Hirsute build. So it's entirely possible this was a problem with the `assert.h` header in a particular Ubuntu release, which (since it wasn't LTS) is probably going to reach its EOL soon, if it hasn't already.

IOW, doing nothing about this may be the correct response.

**Comment by ferdnyc** [comment]
@plugwash 

> I still believe that including standard C headers within C++ namespaces is not something programs can reasonably expect to work correctly.

That's the thing, though... I can't find any evidence that actually happens, under normal circumstances.

The code provided in the OP not only builds fine under Fedora, but it also builds fine under Ubuntu 18.04 using `g++` 7.5.0. And looking at the preprocessor output (`g++ -E ...`), `__assert_fail` is properly wrapped in `extern "C" { }` and not in any namespace.

That's mostly because it  appears to be _my_ include that's generating the code that defines the symbols, not the one from ImageMagick's headers. I don't see any sign of `assert` symbols in the early ImageMagick-owned parts of the preprocessed output (other than `static_assert` which the system headers are littered with). That may have something to do with the fact that the code doesn't make any IM calls, but I have all optimizations turned off, though, so I wouldn't expect anything to be getting dropped by the compiler.

I'm still at a loss to explain exactly what conditions cause this to happen, but in the majority of cases the IM header code really doesn't seem to be the cause of any questionable code getting generated. But then there are these corner cases where it apparently does.

**Comment by plugwash** [comment]
As I understand it it getting the failure requires a combination of several factors

1. The imagemagick headers are the first thing to include assert.h
2. The imagemagick headers are included via Magick++.h
3. Debian's patched version of assert.h is used.
4. Something other than the imagemagick headers tries to make use of assert functionality.

If any of these are not the case the build will succeed.

**Comment by ferdnyc** [comment]
Slight correction, I did some more poking around.

On Fedora, the first `assert.h` include _is_ within `namespace MagickCore { }`. However, it's also inside an `extern "C" { }`, and more importantly it doesn't interfere with the subsequent `#include` from the main file which defines the same symbols in the global namespace.

Building on Ubuntu 18.04, I can't find _any_ sign of `assert.h` in the preprocessor output **other than** the main-source include at the very bottom of the file, which I guess explains why this isn't reproducing for me on that machine either.

I suppose I could be missing some part of ImageMagick that would activate the problem code, or if it's a weird preprocessor-ordering heisenbug. (Though it appears to be reliably reproducible _under the conditions where it occurs_... whatever those might be, exactly.)

**Comment by ferdnyc** [comment]
> As I understand it it getting the failure requires a combination of several factors

AFAICT I have all those boxes ticked. I'm using the exact code from the OP which covers items 1, 2, and 4, and the `assert.h` on both Ubuntu 18.04 and 20.04 (which I've now tried, and also failed to repro) would seem to qualify:
```sh
ubuntu2004% grep -n _ASSERT_H_DECLS /usr/include/assert.h
64:#ifndef _ASSERT_H_DECLS
65:#define _ASSERT_H_DECLS
86:#endif /* Not _ASSERT_H_DECLS */
```
Yet, no can haz errurz:
```sh
ubuntu2004% g++ `pkg-config --cflags ImageMagick++` test.cpp         
ubuntu2004% ./a.out 
a.out: test.cpp:4: int main(): Assertion `1==2' failed.
zsh: abort (core dumped)  ./a.out
```

**Comment by plugwash** [comment]
I just reproduced the issue on Debian sid, then started digging into the code. It looks like a duplicate of this issue was filed and fixed about a fortnight ago.

https://github.com/ImageMagick/ImageMagick6/commit/88543494464bf94ef3430ba33bb34c62ba5050d6#diff-3268ce380c3c96855c30ea8875946bebcb93394c3f6e4b64d18028d73d618316

So that would explain why you can't reproduce it if you are using  imagemagick from upstream git.
